### PR TITLE
perf(api): optimise count retrieval

### PR DESCRIPTION
**Pull request by PJColombo**
### Checklist

- [ ] My change requires a documentation update, and I have done it.
- [x] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description
It optimizes the retrieval of element counts when filters are applied. By obtaining the total number of elements from the stats tables—specifically when filtering by rollup, category, or timestamp—we can utilize data from the overall or daily stats tables.

### CODE REVIEW

File: packages/api/src/utils/count.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  const addressFiltersCount = transactionAddresses?.length ?? 0;
+  const hasRollupAsAddress = !!getRollupFromAddressFilter(transactionAddresses);
+  const reorgedFilterEnabled = !!blockType?.some;
+
+  return (
+    blockNumber ||
+    blockSlot ||
+    reorgedFilterEnabled ||
+    addressFiltersCount > 1 ||
+    (addressFiltersCount === 1 && !hasRollupAsAddress)
```
Comment (xFJA) [inline packages/api/src/utils/count.ts:21]: Maybe we could create const for readability

**Review by xFJA — APPROVED** [review]
Nice work!! 💯

File: packages/api/src/routers/blob/getCount.ts [hunk 2]
```diff
@@ -19,52 +19,74 @@ export async function countBlobs(
...
+            }
+          : undefined,
+      },
+    });
+  }
+
+  const { fromDay, toDay, category, rollup } = buildCountStatsFilters(filters);
 
-  return prisma.blobsOnTransactions.count({
+  if (fromDay || toDay) {
```
Comment (PabloCastellano) [inline packages/api/src/routers/blob/getCount.ts:57]: A comment like this would be helpful

```suggestion
  // Counting elements in postgres can become very slow when there are millions of them in a table, so we take advantage of the calculated data used in charts
  if (fromDay || toDay) {
  ```

File: packages/db/prisma/extensions/stats.ts [hunk 3]
```diff
@@ -32,7 +32,9 @@ export const statsExtension = Prisma.defineExtension((prisma) =>
           });
         },
         populate(dailyDatePeriod?: DatePeriodLike) {
-          const { from, to } = toDailyDatePeriod(dailyDatePeriod);
+          const { from: from = MIN_DATE, to: to = new Date() } = dailyDatePeriod
```
Comment (PabloCastellano) [inline packages/db/prisma/extensions/stats.ts:35]: IMO it was clearer as it was before, when the function did the checks and set the defaults

File: packages/api/test/test-suites/filters.ts [hunk 4]
```diff
@@ -124,12 +124,6 @@ export function runFilterTests(
       });
     });
 
-    it("should return the correct results when filtering by a 'null' rollup", async () => {
```
Comment (PabloCastellano) [inline packages/api/test/test-suites/filters.ts:127]: why did you remove this test case?

File: packages/api/test/tx.test.ts [hunk 5]
```diff
@@ -94,15 +96,18 @@ describe("Transaction router", async () => {
...
+      await ctx.prisma.transactionDailyStats.populate();
 
       const { totalTransactions } = await caller.tx.getCount({});
 
       expect(totalTransactions).toBe(fixtures.canonicalTxs.length);
     });
 
     runFilterTests(async (filters) => {
+      await ctx.prisma.transactionOverallStats.populate();
+      await ctx.prisma.transactionDailyStats.populate();
```
Comment (PabloCastellano) [inline packages/api/test/tx.test.ts:110]: should we move these two lines to beforeAll()?

File: packages/api/src/routers/blob/getCount.ts [hunk 6]
```diff
@@ -19,52 +19,74 @@ export async function countBlobs(
   prisma: BlobscanPrismaClient,
```
Comment (PabloCastellano) [inline packages/api/src/routers/blob/getCount.ts:19]: Since this function is now more complex and we use the precalculated values to improve the performance... can you add a comment to this function explaining that? 

Also adding some test cases for this function would be helpful to check that the returned values didn't change.

File: packages/api/src/routers/tx/getCount.ts [hunk 7]
```diff
@@ -16,19 +16,6 @@ const outputSchema = z.object({
 });
 
 export async function countTxs(prisma: BlobscanPrismaClient, filters: Filters) {
-  if (!hasCustomFilters(filters)) {
```
Comment (PabloCastellano) [inline packages/api/src/routers/tx/getCount.ts:19]: Same as with countBlobs:

- add comment explaining the logic we use now
- add test cases

File: packages/api/src/utils/count.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import { toDailyDatePeriod } from "@blobscan/dayjs";
+
+import { getRollupFromAddressFilter } from "../middlewares/withFilters";
+import type { Filters } from "../middlewares/withFilters";
+
+export function hasToPerformCount({
```
Comment (PabloCastellano) [inline packages/api/src/utils/count.ts:6]:
```suggestion
/* This function determines if the result can be obtained using pre-calculated values from charts or not. */
export function hasToPerformCount({
```

File: packages/dayjs/src/date.ts [hunk 9]
```diff
@@ -51,17 +51,22 @@ export function toDailyDate(
...
 
-  return {
-    from: from ? toDailyDate(from, "startOf").toDate() : MIN_DATE,
-    to: to ? toDailyDate(to, "endOf").toDate() : new Date(),
-  };
+  const datePeriod: DatePeriod = {};
+
+  if (from) {
+    datePeriod.from = toDailyDate(from, "startOf").toDate();
+  }
```
Comment (PabloCastellano) [inline packages/dayjs/src/date.ts:65]:
```suggestion
  } else {
    datePeriod.from = MIN_DATE;
  }
````

File: packages/dayjs/src/date.ts [hunk 10]
```diff
@@ -51,17 +51,22 @@ export function toDailyDate(
...
+
+  if (to) {
+    datePeriod.to = toDailyDate(to, "endOf").toDate();
+  }
```
Comment (PabloCastellano) [inline packages/dayjs/src/date.ts:69]:
```suggestion
  } else {
    datePeriod.to = new Date();
  }
```

File: packages/db/prisma/extensions/stats.ts (hunk 3 above)
Comment (PJColombo) [reply packages/db/prisma/extensions/stats.ts:35]: I encountered some errors while running tests, and it took me a while to trace them back to this function setting default date values.

To prevent similar unexpected side effects in the future, I’d prefer to align this function with the`toDailyDate` logic, focusing solely on date normalization without setting default values.

File: packages/api/src/routers/blob/getCount.ts (hunk 6 above)
Comment (PJColombo) [reply packages/api/src/routers/blob/getCount.ts:28]: Done

File: packages/api/test/tx.test.ts (hunk 5 above)
Comment (PJColombo) [reply packages/api/test/tx.test.ts]: These were removed due to latest tests updates

File: packages/api/test/test-suites/filters.ts (hunk 4 above)
Comment (PJColombo) [reply packages/api/test/test-suites/filters.ts]: I removed it while testing and forgot to re-add it. I’ve just pushed a commit to add it back.
